hal/src/common/dmac/dma_controller.rs

```diff
@@ -0,0 +1,420 @@
+//! # Abstractions to setup and use the DMA controller
+//!
+//! # Initializing
+//!
+//! The DMAC should be initialized using the
+//! [`DmaController::init`](DmaController::init) method. It will consume the
+//! DMAC object generated by the PAC. By default, all four priority levels
+//! will be enabled, but can be selectively enabled/disabled through the
+//! `level_x_enabled` methods.
+//!
+//! # Splitting Channels
+//!
+//! Using the [`DmaController::split`](DmaController::split) method will return
+//! a struct containing handles to individual channels.
+//!
+//! # Releasing the DMAC
+//!
+//! Using the [`DmaController::release`](DmaController::release) method will
+//! deinitialize the DMAC and return the underlying PAC object.
+
+#[cfg(any(feature = "samd11", feature = "samd21"))]
+pub use crate::target_device::dmac::chctrlb::{
+    LVL_A as PriorityLevel, TRIGACT_A as TriggerAction, TRIGSRC_A as TriggerSource,
+};
+
+#[cfg(any(
+    feature = "samd51",
+    feature = "same51",
+    feature = "same53",
+    feature = "same54"
+))]
+pub use crate::target_device::dmac::channel::{
+    chctrla::{
+        BURSTLEN_A as BurstLength, THRESHOLD_A as FifoThreshold, TRIGACT_A as TriggerAction,
+        TRIGSRC_A as TriggerSource,
+    },
+    chprilvl::PRILVL_A as PriorityLevel,
+};
+
+use super::{
+    channel::{new_chan, Channel, Uninitialized},
+    DESCRIPTOR_SECTION, WRITEBACK,
+};
+use crate::target_device::{DMAC, PM};
+
+/// Initialized DMA Controller
+pub struct DmaController {
+    dmac: DMAC,
+}
+
+impl AsMut<DMAC> for DmaController {
```

Comment:
If I'm not mistaken, the implementation would allow users to get a reference to the `DMAC` struct in safe code. That would allow them to make arbitrary changes to the DMA configuration. I would propose adding `unsafe` functions called `dmac` and `dmac_mut` to serve the same purpose. That way users must use `unsafe` to access the `DMAC` directly.